lib/extras/gain_map.cc

```diff
@@ -0,0 +1,225 @@
+class FixedSizeMemoryManager {
+ public:
+  FixedSizeMemoryManager() = default;
+
+  JxlMemoryManager* memory_manager() { return &manager_; }
+
+ private:
+  static void* FixedSizeMemoryManagerAlloc(void* opaque, size_t capacity) {
+    auto manager = static_cast<FixedSizeMemoryManager<N>*>(opaque);
+    if (capacity > N + jxl::memory_manager_internal::kAlias) {
+      return nullptr;
+    }
+    return manager->memory_;
+  }
+  static void FixedSizeMemoryManagerFree(void* opaque, void* pointer) {}
+
+  uint8_t memory_[N + jxl::memory_manager_internal::kAlias];
+  JxlMemoryManager manager_ = {
+      /*opaque=*/this,
+      /*alloc=*/&FixedSizeMemoryManagerAlloc,
+      /*free=*/&FixedSizeMemoryManagerFree,
+  };
+};
+
+}  // namespace
+
+JXL_BOOL JxlGainMapGetBundleSize(const JxlGainMapBundle* map_bundle,
+                                 size_t* bundle_size) {
+  if (map_bundle == nullptr) return JXL_FALSE;
+
+  jxl::ColorEncoding internal_color_encoding;
+  size_t color_encoding_size = 0;
+  size_t extension_bits = 0;
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::CanEncode(internal_color_encoding, &extension_bits,
+                                &color_encoding_size)) {
+      return JXL_FALSE;
+    }
+  }
+
+  *bundle_size =
+      1 +                                     // size of jhgm_version
+      2 +                                     // size_of gain_map_metadata_size
+      map_bundle->gain_map_metadata_size +    // size of gain_map_metadata
+      1 +                                     // size of color_encoding_size
+      jxl::DivCeil(color_encoding_size, 8) +  // size of the color_encoding
+      4 +                                     // size of compressed_icc_size
+      map_bundle->alt_icc_size +              // size of compressed_icc
+      map_bundle->gain_map_size;              // size of gain map
+  return JXL_TRUE;
+}
+
+JXL_BOOL JxlGainMapWriteBundle(const JxlGainMapBundle* map_bundle,
+                               uint8_t* output_buffer,
+                               size_t output_buffer_size,
+                               size_t* bytes_written) {
+  if (map_bundle == nullptr) return JXL_FALSE;
+
+  uint8_t jhgm_version = map_bundle->jhgm_version;
+
+  FixedSizeMemoryManager<sizeof(jxl::ColorEncoding)> memory_manager;
+  jxl::ColorEncoding internal_color_encoding;
+  jxl::BitWriter color_encoding_writer(memory_manager.memory_manager());
+  if (map_bundle->has_color_encoding) {
+    JXL_RETURN_IF_ERROR(
+        internal_color_encoding.FromExternal(map_bundle->color_encoding));
+    if (!jxl::Bundle::Write(internal_color_encoding, &color_encoding_writer,
+                            /*layer=*/0, nullptr)) {
+      return JXL_FALSE;
+    }
+  }
+
+  color_encoding_writer.ZeroPadToByte();
+
+  uint64_t cursor = 0;
+  uint64_t next_cursor = 0;
+
+#define ADVANCE_CURSOR(n)                        \
```

Comment:
I don't like the naming too much (it seems to imply that it advances the cursor by `n`, which it doesn't), but it's not a super strong opinion.

Reply:
how about "SAFE_CURSOR_UPDATE"?